File: 00931. Minimum Falling Path Sum/00931.c
// https://leetcode.com/problems/minimum-falling-path-sum

int getMinUpper(int **m, int row, int col, int colSize) {
    int a = INT_MAX, b = INT_MAX, c = INT_MAX;
    row = row-1; // last row

    if (0 <= col-1) {
        a = m[row][col-1];
    }
    if (0 <= col && col < colSize) {
        b = m[row][col];
    }
    if (col+1 < colSize) {
        c = m[row][col+1];
    }

    int min = fmin(a, b);
    return fmin(c, min);
}

int minFallingPathSum(int **matrix, int rowSize, int *matrixColSize) {
    int min = INT_MAX;
    int colSize=*matrixColSize;

    for (int row=1, col; row<rowSize; ++row) {
        for (col=0; col<colSize; ++col) {
            matrix[row][col] += getMinUpper(matrix, row, col, colSize);
        }
    }
    
    for (int row=rowSize-1, col=0; col<colSize; ++col) {
        if (min > matrix[row][col])
            min = matrix[row][col];
    }

    return min;
}

File: 00931. Minimum Falling Path Sum/00931.go
// https://leetcode.com/problems/minimum-falling-path-sum

func minFallingPathSum(matrix [][]int) int {
    n := len(matrix)
    if n == 1 {
        return matrix[0][0]
    }

    nextRow := func(i, j int) int {
        if j == 0 {
            return min(matrix[i+1][j], matrix[i+1][j+1])
        } else if j == n-1 {
            return min(matrix[i+1][j-1], matrix[i+1][j])
        } else {
            return min(min(matrix[i+1][j-1], matrix[i+1][j]), matrix[i+1][j+1])
        }
    }

    for row := n-2; row > 0; row-- {
        for col := 0; col < n; col++ {
            matrix[row][col] += nextRow(row, col)
        }
    }

    min := 10000
    for col := 0; col < n; col++ {
        if sum := matrix[0][col] + nextRow(0, col); min > sum {
            min = sum
        }
    }
    return min
}
